### ISSUE TITLE
run kube-proxy on master node

### ISSUE THREAD

**Issue by pwais** [kubernetes/kubernetes#9623]
kube-proxy does not currently run on the master node (at least for providers that use salt).  Brendan Burns notes that it might be a good idea (generally) to add this: https://groups.google.com/d/msg/google-containers/PLZonkj9N4g/HjDKEFGP4i0J

I'm opening this issue to:
- query if this is actually a desirable change
- and if so, offer to contribute a patch

I needed this functionality when using a docker registry running inside the k8s cluster as a pod / service.

My draft of the change is here (see bottom; this branch includes some changes not relevant to this issue): https://github.com/GoogleCloudPlatform/kubernetes/compare/master...pwais:pwais_run_proxy_on_master_and_make_priv_nicer_0.18.1

In summary,
- we must modify some of the provider scripts to provide the api_server config to the master (currently missing)
- we must tell salt to install and start the kube-proxy on the master

Issues with this draft:
- I've tested GCE, but not sure about AWS or vagrant.  `make test` passes, but my guess is this config code isn't unit tested
- I might be missing some providers that don't use salt
- The current branch has some unrelated changes

If this is something k8s wants, I'd be happy to help if I can.

**Comment by thockin** [comment]
I'm a little wary of making master-components depend on DNS, it feels a
little bit like a chicken-egg problem.  Can you explain why you need this
for a registry?  I think I am missing a piece.

**Comment by pwais** [comment]
I was using the registry to serve images that k8s would then pull and run.  Without kube-proxy on the master, I was seeing errors when k8s would try to start a pod using an image hosted on a pod / service IP... If this problem sounds weird, perhaps I had something else set up wrong (or was using an old version of k8s when I started working on this).  Brendan's advice (in the linked thread) seemed to indicate kube-proxy on the master was necessary (or at least a good idea generally).

**Comment by piosz** [comment]
Enabling kube-proxy on master node will be useful for Autoscaling purposes where we connect from master node (api-server and controller-manager) to cluster services running on nodes (InfluxDB and Heapster). I'm adding PR which turns it on.

**Comment by roberthbailey** [comment]
The apiserver can already connect to services in the cluster. 

AFAIK, right now the controller manager's only interaction with the cluster is through the apiserver (look at current state, try to make it match desired state). It seems like a major architectural change to have it reach out to things running inside the cluster. 

How do the master components decide which cluster services to talk to? Not all clusters will be running InfluxDB and Heapster.

**Comment by pwais** [comment]
@piosz thanks for that PR!  Looks like pretty much exactly what I tried (and what worked for me as of at least v1.0).  


> The apiserver can already connect to services in the cluster without needing the kube-proxy.

I could not get a private docker registry running in v0.17 (and I believe v1.0.0) w/out this change.  I'd run the registry as a k8s service and then try to start pods that would pull images from the service.  I was able to confirm that the nodes/minions could connect to the service, but the master could not.  This was on GCE.  A lot has changed since I last tested this, though; e.g. #12833

**Comment by piosz** [comment]
@pwais sorry but I no longer need it and as @roberthbailey wrote in https://github.com/kubernetes/kubernetes/pull/13196#issuecomment-135113228 it won't work on hosted master, so I'm closing my PR. If you think it should be merged feel free to convince @thockin and I'd be happy to continue work on it.

**Comment by pwais** [comment]
I guess I'll see if I can reproduce the docker registry service connection issues I saw in 0.17.  Or if somebody can reference evidence of a k8s using images from a cluster-hosted docker registry, that would help me personally and also probably warrant closing the issue altogether ^_^

**Comment by cmluciano** [comment]
kube-proxy now runs by default on the controller with kubeadm.
